### PR TITLE
[AMB-4756]: Implement redesigned Bridge feature

### PR DESCRIPTION
**Pull request by EvgeniyJB**
4756 https://inc4net.atlassian.net/jira/software/c/projects/AMB/boards/6?assignee=60ae0bbd1ee723006c080e4b&selectedIssue=AMB-4756

### CODE REVIEW

File: src/features/bridge/context/Bridge.Context.tsx [hunk 1]
```diff
@@ -1,200 +1,236 @@
...
-  DEFAULT_TOKEN_PAIRS
-} from '../constants';
-import { getBridgeBalance, getBridgePairs } from '@lib';
-import { parseNetworkParams } from '@features/bridge/hooks/bridge/services';
+  DEFAULT_TOKEN_FROM,
+  DEFAULT_TOKEN_TO
+} from '@features/bridge/constants';
+import { useWallet } from '@hooks';
+import { getBridgeConfig } from '@features/bridge/context/helpers';
+import { parsedBridges } from '@features/bridge/context/helpers/parseBridges';
```
Comment (ArturHoncharuk) [inline src/features/bridge/context/Bridge.Context.tsx:16]: src/features/bridge/context/helpers -> src/features/bridge/utils

File: src/features/bridge/context/helpers/getBridgeFee.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  // we need to subtract 1 from the balance to avoid inaccuracies in calculations about the maximum transfer amount
+  const amountToSubtract = '1';
+  const balanceToFee = selectedTokenFrom.isNativeCoin
+    ? // @ts-ignore
+      selectedTokenFrom.balance.sub(
+        parseUnits(amountToSubtract, selectedTokenFrom.decimals)
+      )
+    : selectedTokenFrom.balance;
+
+  const isAmountToFeeLessThenZero = !balanceToFee.gt(0);
```
Comment (ArturHoncharuk) [inline src/features/bridge/context/helpers/getBridgeFee.ts:29]: balanceToFee.isZero()

File: src/features/bridge/hooks/usePendingTransactions.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { useBridgeHistory } from '../../../hooks/query/useBridgeHistory';
+
+export const usePendingTransactions = () => {
+  const { data: transactions } = useBridgeHistory();
+
+  const transactionInPending = transactions.find(
```
Comment (ArturHoncharuk) [inline src/features/bridge/hooks/usePendingTransactions.ts:6]: wrap it with useMemo

File: src/features/bridge/templates/BottomSheetBridgeNetworkSelector/index.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  const { onPressItem, type: destination, selectorType } = props;
+
+  const isFrom = selectorType && destination === 'from';
+
+  const { variables } = useBridgeContextData();
+  const { bridges, fromData, destinationData } = variables;
+
+  const pickerData = isFrom ? fromData : destinationData;
+
+  const { t } = useTranslation();
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgeNetworkSelector/index.tsx:34]: move it to the top of component, above {props}

File: src/features/bridge/templates/BottomSheetBridgeNetworkSelector/index.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+interface BottomSheetChoseNetworksProps {
+  onPressItem: (item: any) => void;
+  selectorType: 'network' | 'token';
+  type?: BridgeSelectorTypes;
+}
+
+export const BottomSheetBridgeNetworkSelector = forwardRef<
+  BottomSheetRef,
+  BottomSheetChoseNetworksProps
+>((props, ref) => {
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgeNetworkSelector/index.tsx:24]: destructure props args

({ onPressItem, type: destination, selectorType}, ref)

File: src/features/bridge/templates/BottomSheetBridgeNetworkSelector/index.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  const { bridges, fromData, destinationData } = variables;
+
+  const pickerData = isFrom ? fromData : destinationData;
+
+  const { t } = useTranslation();
+
+  // @ts-ignore
+  const onClose = () => ref?.current?.dismiss();
+
+  const renderNetworkItem = (network: any) => {
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgeNetworkSelector/index.tsx:39]: wrap it with useCallback + network type is ListRenderItemInfo<_type>

File: src/features/bridge/templates/BottomSheetBridgePreview/components/ErrorTemplate/ErrorTemplate.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import { Text } from '@components//base';
+import { useTranslation } from 'react-i18next';
+import { COLORS } from '@constants/colors';
+import { scale } from '@utils/scaling';
+
+export const ErrorTemplate = () => {
+  const { t } = useTranslation();
+
+  return (
+    <View style={{ alignItems: 'center' }}>
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgePreview/components/ErrorTemplate/ErrorTemplate.tsx:14]: inline styles(

File: src/features/bridge/templates/BottomSheetBridgePreview/components/GeneralPreviewTemplate/GeneralPreviewTemplate.tsx [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+          setProcessingTransaction({
+            ...processingTransaction,
+            eventId: 0,
+            userTo: transaction.to,
+            withdrawTx: transaction.transactionHash
+          });
+          return transaction;
+        });
+      } catch (e) {
+        // ignore
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgePreview/components/GeneralPreviewTemplate/GeneralPreviewTemplate.tsx:62]: do not ignore error block, throw error

File: src/features/bridge/templates/BottomSheetBridgePreview/components/GeneralPreviewTemplate/GeneralPreviewTemplate.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      <PreviewDataTemplate
+        errorBalance={errorBalance}
+        onAcceptPress={onAcceptPress}
+      />
+    );
+  }
+
+  return (
+    <View>
+      <Row justifyContent={'space-between'}>
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgePreview/components/GeneralPreviewTemplate/GeneralPreviewTemplate.tsx:86]: {'space-between'} -. "space-between"

File: src/features/bridge/templates/BottomSheetBridgePreview/components/PreviewData/components/PreviewDataItem.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import { NumberUtils } from '@utils/number';
+import { formatUnits } from 'ethers/lib/utils';
+import { View } from 'react-native';
+import { Row, Spacer, Text } from '@components/base';
+import { TokenLogo } from '@components/modular';
+import { COLORS } from '@constants/colors';
+import { scale } from '@utils/scaling';
+import React from 'react';
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgePreview/components/PreviewData/components/PreviewDataItem.tsx:8]: try to sort imports in such way:

- React
- react-native
- other npm libs
than import local files

File: src/features/bridge/templates/BottomSheetBridgePreview/components/PreviewData/components/PreviewDataItem.tsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    formatUnits(renderItem.crypto.amount, renderItem.crypto.decimals),
+    5
+  );
+
+  const isFirst = index === 0;
+
+  return (
+    <>
+      <View
+        style={{
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgePreview/components/PreviewData/components/PreviewDataItem.tsx:32]: inline style

File: src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/BottomSheetBridgeTransactionPendingHistory.tsx [hunk 12]
```diff
@@ -80,126 +38,24 @@ export const BottomSheetBridgeTransactionPendingHistory = forwardRef<
...
-            color={COLORS.neutral800}
-            style={styles.heading}
-          >
-            {t('bridge.transaction.sending.heading', {
-              tokenFrom: `${formattedAmount} ${transaction?.tokenFrom?.name}`,
-              networkTo:
-                SHORTEN_NETWORK[transaction?.networkTo as keyof typeof NETWORK]
-            })}
-          </Text>
+      <View style={{ paddingHorizontal: scale(24) }}>
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/BottomSheetBridgeTransactionPendingHistory.tsx:41]: inline style

File: src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/components/BridgeTransactionPendingTemplate/BridgeTransactionPendingTemplate.tsx [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        : 'success';
+    },
+    [liveTransactionInformation]
+  );
+
+  return (
+    <View style={styles.container}>
+      <Row
+        justifyContent="space-between"
+        style={{
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/components/BridgeTransactionPendingTemplate/BridgeTransactionPendingTemplate.tsx:54]: inline style

File: src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/components/TransactionLoader/TransactionLoader.tsx [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import { View } from 'react-native';
+import { Spinner } from '@components/base';
+import React from 'react';
+
+export const TransactionLoader = () => (
+  <View
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/components/TransactionLoader/TransactionLoader.tsx:6]: inline style

File: src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/components/TransactionLoader/TransactionLoader.tsx [hunk 15]
```diff
@@ -0,0 +1,15 @@
+import { View } from 'react-native';
+import { Spinner } from '@components/base';
+import React from 'react';
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BottomSheetBridgeTransactionPendingHistory/components/TransactionLoader/TransactionLoader.tsx:3]: sort imports

File: src/features/bridge/templates/BridgeForm/BridgeForm.tsx [hunk 16]
```diff
@@ -13,221 +11,206 @@ import Animated, {
...
 import { DEVICE_HEIGHT } from '@constants/variables';
+import { PrimaryButton } from '@components/modular';
+import { View } from 'react-native';
+import { scale, verticalScale } from '@utils/scaling';
+import { isAndroid } from '@utils/isPlatform';
+import { COLORS } from '@constants/colors';
+import { TokenSelectData } from '@features/bridge/templates/BridgeForm/components/TokenSelectData/TokenSelectData';
+import { InputWithTokenSelect } from '@components/templates';
+import { Tokens } from '@models/Bridge';
+import { parseUnits } from 'ethers/lib/utils';
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BridgeForm/BridgeForm.tsx:27]: sort imports

File: src/features/bridge/templates/BridgeForm/BridgeForm.tsx [hunk 17]
```diff
@@ -13,221 +11,206 @@ import Animated, {
...
+        crypto: {
+          amount: gasFee,
+          decimals: networkNativeToken?.decimals || 18
+        },
+        symbol: networkNativeToken.symbol
+      }
+    ];
+  };
+
+  const goToPreview = async () => {
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BridgeForm/BridgeForm.tsx:128]: useCallback, required to optimise such funcs

File: src/features/bridge/templates/BridgeForm/components/TokenSelectData/TokenSelectData.tsx [hunk 18]
```diff
@@ -0,0 +1,70 @@
+import React, { forwardRef } from 'react';
+import { BottomSheetRef } from '@components/composite';
+import { useBridgeContextData } from '@features/bridge/context';
+import { FlatList, View } from 'react-native';
+import { RenderTokenItem } from '@models/Bridge';
+import { styles } from './styles';
+import { BridgeSelectorItem } from '@components/templates/BridgeSelectorItem';
+import { NumberUtils } from '@utils/number';
+import { formatUnits } from 'ethers/lib/utils';
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BridgeForm/components/TokenSelectData/TokenSelectData.tsx:9]: sort imports

File: src/features/bridge/templates/BridgeHistory/BridgeHistory.Transactions.tsx [hunk 19]
```diff
@@ -11,7 +11,7 @@ import { styles } from './styles';
 import { LocalizedRenderEmpty } from '@components/templates';
 import { Spacer, Text } from '@components/base';
 import { CenteredSpinner } from '@components/composite';
-import { useBridgeHistory } from '@hooks';
+import { useBridgeHistory } from '../../../../hooks/query/useBridgeHistory';
```
Comment (ArturHoncharuk) [inline src/features/bridge/templates/BridgeHistory/BridgeHistory.Transactions.tsx:14]: avoid using relative import, use alias instead

File: src/models/Bridge.ts [hunk 20]
```diff
@@ -26,11 +27,10 @@ export interface BridgePairsModel {
   pairs: [Token, Token][];
   provider: Promise<ethers.providers.JsonRpcProvider | undefined>;
 }
-export interface RenderTokenItem extends Token {
+export interface _RenderTokenItem extends Token {
```
Comment (ArturHoncharuk) [inline src/models/Bridge.ts:30]: seems like bad naming for interface, try to use TokenItemArgs

File: src/models/dtos/Bridge.ts [hunk 21]
```diff
@@ -20,8 +20,5 @@ export interface BridgeTransactionHistoryDTO {
   withdrawTx: string;
   timestampStart: number;
   transferFinishTxHash: string;
-}
-
-export interface BridgeTransactionDTO extends BridgeTransactionHistoryDTO {
-  loading?: boolean;
+  wait: () => Promise<any>;
```
Comment (ArturHoncharuk) [inline src/models/dtos/Bridge.ts:23]: use unknown instead of any

File: src/screens/Bridge/Bridge.tsx [hunk 22]
```diff
@@ -1,42 +1,52 @@
...
+  }
+
   return (
     <SafeAreaView style={styles.container}>
       <Header
-        onBackPress={() => {
-          setDefaultBridgeData();
-          navigation.goBack();
-        }}
+        onBackPress={() => navigation.goBack()}
```
Comment (ArturHoncharuk) [inline src/screens/Bridge/Bridge.tsx:49]: onBackPress={() => navigation.goBack()} -> onBackPress={navigation.goBack}

File: src/features/bridge/context/helpers/getBridgeFee.ts (hunk 2 above)
Comment (EvgeniyJB) [reply src/features/bridge/context/helpers/getBridgeFee.ts:29]: balanceToFee.isZero() return true only for 0. !balanceToFee.gt(0) return true if amount less then 0

**Review by ArturHoncharuk — APPROVED** [review]

**Pull request merged**